Formula/lean.rb

```diff
@@ -29,6 +29,11 @@ class Lean < Formula
     sha256 cellar: :any_skip_relocation, x86_64_linux:   "47716dec0f9e482dd8a44694c4916cc86d104dcc65f50f4fc8d81e0f2d366ef6"
   end
 
+  # Lean 3 is now at end of life.
+  # The `elan-init` formula provides a Lean installation manager
+  # which continues to support Lean 3 users, but also provides Lean 4.
+  disable! date: "2024-06-28", because: :deprecated_upstream
```

Comment:
Why not update this to Lean 4 then?

Reply:
Speaking as a Lean developer, it does not make sense to package a single version of a compiler that has no compatibility guarantees between any two versions. Packaging only elan is the correct choice.

Reply:
But people submitted this formula so clearly there is demand?

Reply:
It should never have been submitted / accepted in the first place. It is not viable to use a fixed version of Lean 3 or Lean 4. All installation instructions ask users to install the `elan` tool (provided by the `elan-init` formula in Homebrew), and to allow it to automatically configure the correct compiler version for each project.

Ideally, we would disable this formula, and eventually rename the `elan-init` formula simply to `lean`, to help avoid confusion. But any formula that installs a fixed version of the compiler without providing `elan` is just going to result in confused users unable to compile projects, or coming to us with requests for help.

Reply:
I have to admit I'm confused here. I promise (as a [Lean developer](https://lean-fro.org/people/)) that people using the formula added in #17224 are getting something they do not want. All of the Lean community's installation instructions tell users to install `elan`. We would prefer not to have to help new users undo the difficulties they've arrived at by installing the Homebrew `lean` formula. The availability of the Homebrew `lean` formula is one of the last remaining places where people are encouraged to install `lean` without `elan`. Perhaps as an analogy it would be like giving someone a Homebrew install script that disabled the `update` functionality, so they would be forever stuck on that version.

It is true that Lean 3.51 is an open source tool. However it is at end of life, the repository is not accepting further commits, and most major projects have moved to Lean 4.

In practice in the Lean ecosystem, every project uses a slightly different version of Lean (whether it is Lean 3 or Lean 4), and updates rapidly. The `elan-init` formula provides a tool that automatically selects and uses the correct version of Lean for each project. The `lean` formula provides something that could not actually compile any of the significant projects in Lean today.

Does that make sense? I'm happy to explain in more detail, or cite sources as needed. :-)

Reply:
I'm not sure what you mean by "why not Lean 4". `elan` supports both Lean 3 and Lean 4, and will provide whichever version of either is required for a specific project.

There is no single "Lean 4" at the moment, and all projects rely on nightly releases. Thus you need `elan` to provide the relevant version for each project. That is, just switching from Lean 3 to Lean 4 does not resolve the problem that we shouldn't be providing anyone a fixed version that is not managed by `elan`. (Monthly official releases will start real-soon-now, but still I hope homebrew can supply `elan` rather than one fixed monthly release.)

I'm sorry, I don't know enough of Homebrew's organisational structure to suggest a solution for how to prevent PRs being merged.

Reply:
Given there's no way it seems to get a stable Lean 4 release: I'm fine with this being disabled.